### PR TITLE
Fixes #13987: Massive performance penalty in policy generation due to invalid usage of StringTemplate

### PR DESCRIPTION
**Pull request by ncharles**
https://issues.rudder.io/issues/13987

### CODE REVIEW

File: rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala [hunk 1]
```diff
@@ -72,8 +72,32 @@ final case class STVariable(
...
+
+  def clearCache(): Unit = this.synchronized {
+    cache = Map()
+  }
+
+  def getTemplateFromContent(content: String): Box[StringTemplate] = this.synchronized {
+    cache.get(content) match {
+      case Some(template) => Full(template.getInstanceOf())
+      case _ =>
+        val template = new StringTemplate(content, classOf[NormationAmpersandTemplateLexer])
```
Comment (fanf) [inline rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala]: I believe that can throw exception, no ? If so, can you wrapp it in `try/catch ex => Failure...`. And the same for `getInstanceOf` if needed.

File: rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala [hunk 2]
```diff
@@ -84,9 +108,7 @@ class FillTemplatesService extends Loggable {
   def fill(templateName: String, content: String, variables: Seq[STVariable]): Box[String] = {
     for {
       template <- try {
-                    //string template does not allows "." in path name, so we are force to use a templateGroup by policy template (versions have . in them)
-                    val template = new StringTemplate(content, classOf[NormationAmpersandTemplateLexer])
-                    Full(template)
+                    getTemplateFromContent(content)
```
Comment (fanf) [inline rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala]: no more need for try/catch here once `getTemplateFromContent` catch exception.

File: rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala (hunk 1 above)
Comment (ncharles) [reply rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala]: you are totally right

**Comment by ncharles** [comment]
Commit modified

File: rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala [hunk 3]
```diff
@@ -72,8 +72,38 @@ final case class STVariable(
...
+   *
+   */
+  private[this] var cache = Map[String,StringTemplate]()
+
+
+  def clearCache(): Unit = this.synchronized {
+    cache = Map()
+  }
+
+  def getTemplateFromContent(templateName: String,content: String): Box[StringTemplate] = this.synchronized {
```
Comment (ncharles) [inline rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala]: i doubled checked - the templateName is indeed the destination of the file, note its source. Which is pretty confusing

Comment (ncharles) [reply rudder-templates/src/main/scala/com/normation/templates/FillTemplatesService.scala]: i dugged into the detail, to see how much ram it took to store the parsed stringtemplate
I have less than 1MB of antlr related class in the heapdump, which is peanuts compared to anything else. Granted, it contains a lot of java instance, but including ALL java instances, I barely reach 200MB (100 MB being Object, 65MB being String)
so i'm pretty convinced the memory usage is not problematic there (at least, we have easier battle to fight to save memory)

**Pull request merged**